services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/switches/SwitchDeleteSpec.groovy

```diff
@@ -1,281 +1,194 @@
+            Wrappers.wait(discoveryTimeout + WAIT_OFFSET) {
+                def links = northbound.getAllLinks()
+                swIsls.each { assert islUtils.getIslInfo(links, it).get().state == IslChangeType.DISCOVERED }
+            }
         }
-        northbound.deleteFlow(flow.id)
+
+        where:
+        [linkType, actionToFix] << [["active", "Unplug and remove them first"], ["inactive", "Remove them first"]]
     }
 
-    def "Cannot delete inactive switch with flow"() {
+    @Unroll
+    def "Unable to delete an inactive switch with a #flowType flow assigned"() {
         requireProfiles("virtual")
 
-        given: "An active switch with flow"
-        def (Switch srcSwitch, Switch dstSwitch) = topology.getActiveSwitches()[0..1]
-        def flow = flowHelper.randomFlow(srcSwitch, dstSwitch)
-        northbound.addFlow(flow)
-        Wrappers.wait(WAIT_OFFSET) { northbound.getFlowStatus(flow.id).status == FlowState.UP }
+        given: "A flow going through a switch"
+        flowHelper.addFlow(flow)
 
-        when: "Deactivate this switch"
-        lockKeeper.knockoutSwitch(srcSwitch.getDpId())
-        Wrappers.wait(WAIT_OFFSET) {
-            !northbound.activeSwitches.any { it.switchId == srcSwitch.dpId}
-        }
+        when: "Deactivate the switch"
+        lockKeeper.knockoutSwitch(flow.source.datapath)
+        Wrappers.wait(WAIT_OFFSET) { assert !northbound.activeSwitches.any { it.switchId == flow.source.datapath } }
 
-        and: "Try to delete inactive switch with flow"
-        northbound.deleteSwitch(srcSwitch.getDpId(), false)
+        and: "Try to delete the switch"
+        northbound.deleteSwitch(flow.source.datapath, false)
 
-        then: "Got 400 BadRequest because switch has flow"
+        then: "Got 400 BadRequest error because the switch has the flow assigned"
         def exc = thrown(HttpClientErrorException)
         exc.rawStatusCode == 400
-        exc.responseBodyAsString.matches(".*Switch '$srcSwitch.dpId' has \\d+ assigned flows.*")
+        exc.responseBodyAsString.matches(".*Switch '${flow.source.datapath}' has \\d+ assigned flows.*")
 
-        and: "Cleanup: Activate the switch back and remove flow"
-        lockKeeper.reviveSwitch(srcSwitch.getDpId())
-        Wrappers.wait(WAIT_OFFSET) {
-            northbound.activeSwitches.any { it.switchId == srcSwitch.dpId }
-        }
-        northbound.deleteFlow(flow.id)
+        and: "Cleanup: activate the switch back and remove the flow"
+        lockKeeper.reviveSwitch(flow.source.datapath)
+        Wrappers.wait(WAIT_OFFSET) { assert northbound.activeSwitches.any { it.switchId == flow.source.datapath } }
+        flowHelper.deleteFlow(flow.id)
+
+        where:
+        [flowType, flow] << [["single-switch", getFlowHelper().singleSwitchFlow(getTopology().getActiveSwitches()[0])],
+                             ["casual", getFlowHelper().randomFlow(*getTopology().getActiveSwitches()[0..1])]]
```

Comment:
Looking at this, I feel like it makes sense to try to delete an intermediate switch, which is not explicitly mentioned as source or destination of certain flow, but is used in path. It looks like a separate valid case for me. This may be out of scope for this pr though

Reply:
Let's have a separate PR for it. I added this test case to my 'todo' list.